Enable full train_step tracing and customizable dist graph expansion

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #97577
* __->__ #97416
* #97534
* #97533

This commit adds an entry point for full `train_step` tracing and
expansion. Model forward, backwrd, and optimizer step will be included
in one graph. DTensor expansion will be applied on top to insert
collective communications. Users can also provide an `Override`
implementation to skip non-traceable submodules and directly install
submodule logic to the  DTensor-expanded graph by inserting `fx.Nodes`.

Differential Revision: [D44325177](https://our.internmc.facebook.com/intern/diff/D44325177)

mrshenli committed:
[WIP] Enable full train_step tracing and customizable dist graph expansion

[ghstack-poisoned]

mrshenli commented on torch/_meta_registrations.py:
```diff
@@ -478,7 +478,7 @@ def meta_mm(a, b):
     N, M1 = a.shape
     M2, P = b.shape
-    check(M1 == M2, lambda: "a and b must have same reduction dim")
+    check(M1 == M2, lambda: f"a and b must have same reduction dim, but got [{N}, {M1}] X [{M2}, {P}].")
```
Will extract this change to a separate PR

mrshenli commented on torch/_prims_common/__init__.py:
```diff
@@ -1063,14 +1063,17 @@ def is_weakly_lesser_type(a: type, b: type) -> bool:
     The comparison is determined by the following type ordering: bool, int, float, complex.
     """
     ordered_types = (
+        torch.SymBool,
```
Will extract this change to a separate PR

mrshenli commented on test/distributed/_spmd/test_tracing.py:
```diff
@@ -206,10 +220,14 @@ def _test_trace_replicate(self, model: nn.Module, x, *args, **kwargs):
             deepcopy(model),
             schema=Schema(
-                mesh=DeviceMesh(self.device_type, torch.arange(self.world_size)),
+                mesh=DeviceMesh(
```
will extract format changes to a separate PR

mrshenli commented on torch/distributed/_spmd/api.py:
```diff
@@ -51,3 +76,209 @@ def forward(self, *args: Tuple[object], **kwargs: Dict[str, object]) -> object:
+
+            # FIXME(@mrshenli): debug only
+            if dist.get_rank() == 0:
+                gm.graph.print_tabular()
```
adding optimizer into the graph does not seem to trigger DTensor partial->replicate allreduce. Will get back to this later. 

```
opcode         name          target                   args                             kwargs                          
-------------  ------------  -----------------------  -------------------------------  ---------------------------------------------------------------------------
---------------------------------------------------------------------------                                                                                       
placeholder    arg0_1        arg0_1                   ()                               {}
placeholder    arg0_2        arg0_2                   ()                               {}
placeholder    arg0_3        arg0_3                   ()                               {}
placeholder    arg2_1        arg2_1                   ()                               {}
placeholder    arg2_2        arg2_2                   ()                               {}
placeholder    arg2_3        arg2_3                   ()                               {}
placeholder    arg2_4        arg2_4                   ()                               {}
placeholder    arg2_5        arg2_5                   ()                               {}
placeholder    arg2_6        arg2_6                   ()                               {}
placeholder    arg3_1        arg3_1                   ()                               {}
placeholder    arg3_2        arg3_2                   ()                               {}
call_function  t_4           aten.t.default           (arg3_1,)                        {}
call_function  addmm_1       aten.addmm.default       (arg3_2, arg0_3, t_4)            {}
call_function  sum_3         aten.sum.default         (addmm_1,)                       {}
call_function  ones_like_1   aten.ones_like.default   (sum_3,)                         {'dtype': torch.float32, 'layout': torch.strided, 'device': device(type='cu
da', index=0), 'pin_memory': False, 'memory_format': torch.preserve_format}                                                                                       
call_function  expand_1      aten.expand.default      (ones_like_1, [4, 10])           {}
call_function  t_5           aten.t.default           (expand_1,)                      {}                              
call_function  slice_1       aten.slice.Tensor        (t_5, 1, 0, 2)                   {}
call_function  clone         aten.clone.default       (slice_1,)                       {}
call_function  mm_1          aten.mm.default          (clone, arg0_3)                  {}
call_function  t_6           aten.t.default           (mm_1,)                          {}
call_function  sum_4         aten.sum.dim_IntList     (expand_1, [0], True)            {}
call_function  view_1        aten.view.default        (sum_4, [10])                    {}
call_function  detach_4      aten.detach.default      (view_1,)                        {}
call_function  detach_5      aten.detach.default      (detach_4,)                      {}
call_function  detach_6      aten.detach.default      (detach_5,)                      {}
call_function  detach_7      aten.detach.default      (detach_6,)                      {}
call_function  t_7           aten.t.default           (t_6,)                           {}
call_function  detach_8      aten.detach.default      (t_7,)                           {}
call_function  detach_9      aten.detach.default      (detach_8,)                      {}
call_function  detach_10     aten.detach.default      (detach_9,)                      {}
call_function  detach_11     aten.detach.default      (detach_10,)                     {}
call_function  add_          aten.add_.Tensor         (arg2_1, 1)                      {}
call_function  mul_          aten.mul_.Tensor         (arg2_2, 0.9)                    {}
call_function  add__1        aten.add_.Tensor         (mul_, detach_11)                {'alpha': 0.09999999999999998}
call_function  mul__1        aten.mul_.Tensor         (arg2_3, 0.999)                  {}
call_function  addcmul_      aten.addcmul_.default    (mul__1, detach_11, detach_11)   {'value': 0.0010000000000000009}
call_function  pow_5         aten.pow.Scalar          (0.9, add_)                      {}
call_function  rsub_4        aten.rsub.Scalar         (pow_5, 1)                       {}
call_function  pow_6         aten.pow.Scalar          (0.999, add_)                    {}
call_function  rsub_5        aten.rsub.Scalar         (pow_6, 1)                       {}
call_function  reciprocal_4  aten.reciprocal.default  (rsub_4,)                        {}
call_function  mul_6         aten.mul.Tensor          (reciprocal_4, 0.01)             {}
call_function  neg_2         aten.neg.default         (mul_6,)                         {}
call_function  sqrt_4        aten.sqrt.default        (rsub_5,)                        {}
call_function  sqrt_5        aten.sqrt.default        (addcmul_,)                      {}
call_function  mul_7         aten.mul.Tensor          (sqrt_4, neg_2)                  {}
call_function  div_2         aten.div.Tensor          (sqrt_5, mul_7)                  {}
call_function  reciprocal_5  aten.reciprocal.default  (neg_2,)                         {}
call_function  mul_8         aten.mul.Tensor          (reciprocal_5, 1e-08)            {}
call_function  add__2        aten.add_.Tensor         (div_2, mul_8)                   {}
call_function  addcdiv_      aten.addcdiv_.default    (arg3_1, add__1, add__2)         {}
call_function  add__3        aten.add_.Tensor         (arg2_4, 1)                      {}
call_function  mul__2        aten.mul_.Tensor         (arg2_5, 0.9)                    {}
call_function  add__4        aten.add_.Tensor         (mul__2, detach_7)               {'alpha': 0.09999999999999998}
call_function  mul__3        aten.mul_.Tensor         (arg2_6, 0.999)                  {}
call_function  addcmul__1    aten.addcmul_.default    (mul__3, detach_7, detach_7)     {'value': 0.0010000000000000009}
call_function  pow_7         aten.pow.Scalar          (0.9, add__3)                    {}
call_function  rsub_6        aten.rsub.Scalar         (pow_7, 1)                       {}
call_function  pow_8         aten.pow.Scalar          (0.999, add__3)                  {}
call_function  rsub_7        aten.rsub.Scalar         (pow_8, 1)                       {}
call_function  reciprocal_6  aten.reciprocal.default  (rsub_6,)                        {}
call_function  mul_9         aten.mul.Tensor          (reciprocal_6, 0.01)             {}
call_function  neg_3         aten.neg.default         (mul_9,)                         {}
call_function  sqrt_6        aten.sqrt.default        (rsub_7,)                        {}
call_function  sqrt_7        aten.sqrt.default        (addcmul__1,)                    {}
call_function  mul_10        aten.mul.Tensor          (sqrt_6, neg_3)                  {}
call_function  div_3         aten.div.Tensor          (sqrt_7, mul_10)                 {}
call_function  reciprocal_7  aten.reciprocal.default  (neg_3,)                         {}
call_function  mul_11        aten.mul.Tensor          (reciprocal_7, 1e-08)            {}
call_function  add__5        aten.add_.Tensor         (div_3, mul_11)                  {}
call_function  addcdiv__1    aten.addcdiv_.default    (arg3_2, add__4, add__5)         {}
output         output        output                   ([None, addcdiv_, addcdiv__1],)  {}
```

mrshenli replied:
update: when using SGD, allreduce can be correctly triggered, but not Adam

mrshenli committed:
Update on "[WIP] Enable full train_step tracing and customizable dist graph expansion"




[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/).

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44325177).

mrshenli commented on torch/distributed/_spmd/distribute.py:
```diff
@@ -437,9 +437,14 @@ def _convert_to_distributed(
-            node_replacements[node] = _get_dtensor_dispatch_graph(
-                node, node_to_obj
-            )
+            if not node.target.__name__.rpartition('.')[0][-1] == "_":
```
I also tried using `torch.no_grad()` to wrap those inplace ops with running DTensor dispatch, which avoids the above autograd error, but DTensor's dispatch graph contains sth like below:

```
get_attr       _tensor_constant0    _tensor_constant0         ()                                      {}                                                          
call_function  transpose            aten.transpose.int        (_tensor_constant0, -1, 0)              {}
```

And it later triggered other errors. Does this ring a bell to you?

```
RuntimeError: Node _tensor_constant0 target _tensor_constant0 references nonexistent attribute _tensor_constant0 of 
```

wanchaol replied:
hmmm I don't know what happened there. Is this only happen when expanding the optimizer graph? I think the current `convert_to_distributed` logic is a bit too messy and hard to maintain. 

Do you think we should do the dtensor expansion in follow up PRs? Hopefully in the meanwhile I can get a new version of expansion in the next few days and we can discuss if we should directly integrate with that.

mrshenli replied:
> hmmm I don't know what happened there. Is this only happen when expanding the optimizer graph? I think the current convert_to_distributed logic is a bit too messy and hard to maintain.

Yep, only happens when including optimizers. `_allow_partial=False` works correctly when we return gradients

> Do you think we should do the dtensor expansion in follow up PRs? Hopefully in the meanwhile I can get a new version of expansion in the next few days and we can discuss if we should directly integrate with that.

Yep, that's the plan. I prefer to land this PR sooner to unblock @yifuwang. And we can work on fixing inplace + DTensor expansion in parallel.

mrshenli committed:
Update on "[WIP] Enable full train_step tracing and customizable dist graph expansion"


Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D44325177](https://our.internmc.facebook.com/intern/diff/D44325177)

[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44325177).

mrshenli commented on torch/distributed/_spmd/api.py:
```diff
@@ -51,3 +76,208 @@ def forward(self, *args: Tuple[object], **kwargs: Dict[str, object]) -> object:
+            # and use DTensor to expand operators, where DTensor's shape is the
+            # global shape.
+            with _enable_compile():
+                gm = make_fx(
```
If I add `functionalize`, I got the following error:

```
  File "/fsx/users/shenli/pytorch/torch/distributed/_spmd/api.py", line 256, in wrapper                                                                           
    gm = functionalize(make_fx(                                                  
  File "/fsx/users/shenli/pytorch/torch/_functorch/vmap.py", line 39, in fn                                                                                       
    return f(*args, **kwargs)                                                    
  File "/fsx/users/shenli/pytorch/torch/_functorch/eager_transforms.py", line 1600, in wrapped                                                                    
    func_outputs = func(*func_args, **func_kwargs)                               
  File "/fsx/users/shenli/pytorch/torch/fx/experimental/proxy_tensor.py", line 715, in wrapped                                                                    
    args = pytree.tree_map(wrap_fn_map[tracing_mode], args)                      
  File "/fsx/users/shenli/pytorch/torch/utils/_pytree.py", line 196, in tree_map                                                                                  
    return tree_unflatten([fn(i) for i in flat_args], spec)                                                                                                       
  File "/fsx/users/shenli/pytorch/torch/utils/_pytree.py", line 196, in <listcomp>                                                                                
    return tree_unflatten([fn(i) for i in flat_args], spec)                                                                                                       
  File "/fsx/users/shenli/pytorch/torch/fx/experimental/proxy_tensor.py", line 704, in wrap_fake                                                                  
    return fake_tensor_mode.from_tensor(x, source=source)  # type: ignore[attr-defined]                                                                           
  File "/fsx/users/shenli/pytorch/torch/_subclasses/fake_tensor.py", line 1398, in from_tensor                                                                    
    return self.fake_tensor_converter(                                                                                                                            
  File "/fsx/users/shenli/pytorch/torch/_subclasses/fake_tensor.py", line 314, in __call__                                                                        
    return self.from_real_tensor(                                                                                                                                 
  File "/fsx/users/shenli/pytorch/torch/_subclasses/fake_tensor.py", line 280, in from_real_tensor                                                                
    raise UnsupportedFakeTensorException("meta converter nyi")                   
torch._subclasses.fake_tensor.UnsupportedFakeTensorException: meta converter nyi  
```

cc @wconstab

mrshenli committed:
Update on "[WIP] Enable full train_step tracing and customizable dist graph expansion"


Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D44325177](https://our.internmc.facebook.com/intern/diff/D44325177)

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Enable full train_step tracing and customizable dist graph expansion"


Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D44325177](https://our.internmc.facebook.com/intern/diff/D44325177)

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Enable full train_step tracing and customizable dist graph expansion"


Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D44325177](https://our.internmc.facebook.com/intern/diff/D44325177)

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Enable full train_step tracing and customizable dist graph expansion"


Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D44325177](https://our.internmc.facebook.com/intern/diff/D44325177)

[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44325177).

mrshenli commented on torch/distributed/_spmd/api.py:
```diff
@@ -51,3 +76,208 @@ def forward(self, *args: Tuple[object], **kwargs: Dict[str, object]) -> object:
+        torch._utils.is_compiling.__code__ = orig_is_compiling_code
+
+
+def compile(override_map: Optional[Dict[Type[Any], Override]] = None):
```
we might have a few options

1. `torch.compile` has an `options` argument which is supposed to pass additional information to backends. 
2. If the type we need does not fit in `options`, we can also just provide that when creating dist backend, given that the backend in `torch.compile` is a Callable. Can be sth like below

```python
class DistBackend:
  def __init__(self, whatever_args_we_need):
    pass

  def __call__(self):
    pass

@torch.compile(backend=DistBackend(whatever_args_we_need))
def user_func():
  pass
```

mrshenli commented on torch/distributed/_spmd/api.py:
```diff
@@ -51,3 +76,208 @@ def forward(self, *args: Tuple[object], **kwargs: Dict[str, object]) -> object:
+            # FIXME(@mrshenli): let make_fx ignore nn.Module in args? or it should
+            # eliminate unused placeholders?
+            # FIXME: Using symbolic tracing to work around. Otherwise it hits
+            # shape mismatch error, as we use local inputs to trace local graph
```
Yep, I tried that but does not seem to work. Maybe I didn't do it correctly. IIUC, the problem I am hitting is that:

*  if I use fake tensor with global batch size, operators like `expand` will embed/hardcode the fake tensor shape as its argument in the graph. But during local execution (before DTensor expansion), it is actually one shard of that, so there will be shape mismatch with the hardcoded one. That expansion operation is from the loss function (e.g., `sum()`). I think we didn't hit this problem in SPMD ([here](https://github.com/pytorch/pytorch/blob/17567e5b29a77dee137dfc365e9ed3f1cf358950/torch/distributed/_spmd/distribute.py#L612)) because SPMD does not include the loss function.
*  If I use local batch size, then the expand operator will hardcode local batch size in the graph. Later during DTensor expansion, the FakeTensor wrapping DTensor will have the global batch size, and hence mismatch.

Not 100% sure if the above understanding is correct.

mrshenli commented on torch/distributed/_spmd/api.py:
```diff
@@ -51,3 +76,208 @@ def forward(self, *args: Tuple[object], **kwargs: Dict[str, object]) -> object:
+            # and use DTensor to expand operators, where DTensor's shape is the
+            # global shape.
+            with _enable_compile():
+                gm = make_fx(
```
> so make_fx would give us the whole graph of fwd + bwd + optim graph?

Yep

> Having the information about what's parameters, what's parameter grads gonna help us with data parallelism.

We can tell what are parameters and optimizer states, as those are explicitly feed into make_fx by us. If necessary, we should also be able to tell what are gradients, as `aot_module/function` does the same? Curious, any reason we need to identify gradients besides parameters and opt states?

mrshenli committed:
Update on "Enable full train_step tracing and customizable dist graph expansion"


This commit adds an entry point for full `train_step` tracing and
expansion. Model forward, backwrd, and optimizer step will be included
in one graph. DTensor expansion will be applied on top to insert
collective communications. Users can also provide an `Override`
implementation to skip non-traceable submodules and directly install
submodule logic to the  DTensor-expanded graph by inserting `fx.Nodes`.

Differential Revision: [D44325177](https://our.internmc.facebook.com/intern/diff/D44325177)

[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44325177).

wanchaol reviewed (approved):
lgtm, thanks!

Let's get this landed first and we can figure out the follow ups about fwd/bwd graphs, and how to do dtensor expand to resolve optimizer issue, we can probably turn on the `allow_partial` to restore the old behavior.

mrshenli commented on torch/distributed/_spmd/api.py:
```diff
@@ -53,3 +76,253 @@ def forward(
+                ):
+                    ret = func(*args, **kwargs)
+                    # make sure updated parameters are returned
+                    return ret, list(mod.parameters())
```
discussed offline, this is to avoid DCE removing nodes not connected to output, as `_convert_to_distributed` still calls into lint and DCE.

mrshenli commented on torch/distributed/_spmd/api.py:
```diff
@@ -53,3 +76,253 @@ def forward(
+        inps.append(p)
+        schemas.append(replicate_schema)
+
+    return _convert_to_distributed(gm, inps, schemas, _allow_partial=False)
```
I assume `_allow_partial = False` was suppose to trigger allreduce manually, but only when the partial tensors are in the graph output. In this case, gradients has been consumed by optimizer ops before reaching output. So even if we include grads in output and force `_allow_partial=False`, the result won't be correct as those allreduce will happen after opt.step?

wanchaol replied:
I see, yeah this gonna be tricky as we directly use `make_fx`... nvm then, I think if we switch to `aot_function` maybe we should still preserve the old behavior, but we can do it in follow up PRs.

mrshenli commented on torch/distributed/_spmd/distribute.py:
```diff
@@ -445,9 +445,16 @@ def _convert_to_distributed(
-            node_replacements[node] = _get_dtensor_dispatch_graph(
-                node, node_to_obj
-            )
+            if not node.target.__name__.rpartition('.')[0][-1] == "_":
```
thanks! Let me update :)

awgu commented on torch/nn/utils/_named_member_accessor.py:
```diff
@@ -82,6 +82,29 @@ def swap_tensor(
+
+    orig_submodule = module._modules[name]
+    if not isinstance(orig_submodule, torch.nn.Module):
+        raise TypeError(f"{name} attribute is not an instance of torch.nn.Module")
```
This branch seems to indicate `nn.Module` itself has inconsistent state, which can only happen if a user abuses the private attributes.

awgu commented on torch/nn/utils/_named_member_accessor.py:
```diff
@@ -128,6 +151,15 @@ def get_submodule(self, name: str) -> "torch.nn.Module":
             self.memo[name] = submodule
             return submodule
 
+    def swap_submodule(self, name: str, value: "torch.nn.Module") -> "torch.nn.Module":
```
nit: Do we want to differentiate the argument name `name` here from the `name` above in the top-level `swap_submodule`? Here, the `name` can be like an FQN (with `'.'`), whereas above, it must be only the unprefixed name.

Maybe call this one `path` if you already refer to it as "path" in the comment?

awgu replied:
Ah, I see that this pattern is consistent with the rest of the file.

mrshenli replied:
good point, let me update

mrshenli committed:
Update on "Enable full train_step tracing and customizable dist graph expansion"


This commit adds an entry point for full `train_step` tracing and
expansion. Model forward, backwrd, and optimizer step will be included
in one graph. DTensor expansion will be applied on top to insert
collective communications. Users can also provide an `Override`
implementation to skip non-traceable submodules and directly install
submodule logic to the  DTensor-expanded graph by inserting `fx.Nodes`.

Differential Revision: [D44325177](https://our.internmc.facebook.com/intern/diff/D44325177)

[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44325177).

mrshenli commented on torch/nn/utils/_named_member_accessor.py:
```diff
@@ -82,6 +82,29 @@ def swap_tensor(
+    if name not in module._modules:
+        raise KeyError(f"submodule {name} does not exist")
+
+    orig_submodule = module._modules[name]
```
no strong reasoning here, just trying to follow the existing convention in `set_tensor` and `swap_tensor` in this file.

mrshenli committed:
Update on "Enable full train_step tracing and customizable dist graph expansion"


This commit adds an entry point for full `train_step` tracing and
expansion. Model forward, backwrd, and optimizer step will be included
in one graph. DTensor expansion will be applied on top to insert
collective communications. Users can also provide an `Override`
implementation to skip non-traceable submodules and directly install
submodule logic to the  DTensor-expanded graph by inserting `fx.Nodes`.

Differential Revision: [D44325177](https://our.internmc.facebook.com/intern/diff/D44325177)

[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44325177).

wconstab commented on torch/distributed/_spmd/api.py:
```diff
@@ -53,3 +82,257 @@ def forward(
+        elif isinstance(a, nn.Module) or isinstance(a, torch.optim.Optimizer):
+            # nn.Module or optimizer placeholder is captured by make_fx but
+            # never used in the graph
+            inps.append(torch.empty(0))
```
did you try an alternative way of doing expansion?  not sure if i got it right but i was thinking 
1) here we can just mutate the input list of args, ... params  (or, return new lists/dicts), where we apply DTensor wrapping to the individual items in these lists
2) then we just return from this function instead of calling _convert_to_distributed
3a) we call _dtensor_expand before  `gm = make_fx(`, so the tracing also triggers dtensor expand automatically
3b) or-- if you want it separate, you just leave _dtensor_expand after first ` gm = make_fx(` and then put another ` gm = make_fx(gm` after _dtensor_expand

The main reason i asked is to simplify code structure, and remove maintenence burden.

mrshenli replied:
Hey @wconstab, thanks for the suggestion! Yep, that's the first thing we tried when developing SPMD. But at that time `make_fx` does not support nested tensor subclass. Not sure if this is already fixed. I can give it try.

wconstab replied:
i was asking @bdhirsh about this yesterday.  He was kind of unsure if it would work or if you would run into an error where basically the mode key for fake tracing was running before the key for the subclass.  If that's still happening, then we should be able to fix it by coming up with a solution for properly ordering the keys.

mrshenli committed:
Update on "Enable full train_step tracing and customizable dist graph expansion"


This commit adds an entry point for full `train_step` tracing and
expansion. Model forward, backwrd, and optimizer step will be included
in one graph. DTensor expansion will be applied on top to insert
collective communications. Users can also provide an `Override`
implementation to skip non-traceable submodules and directly install
submodule logic to the  DTensor-expanded graph by inserting `fx.Nodes`.

Differential Revision: [D44325177](https://our.internmc.facebook.com/intern/diff/D44325177)

[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44325177).

mrshenli commented:
@pytorchbot merge -g

mrshenli commented:
@pytorchbot merge

XuehaiPan commented on torch/nn/utils/_named_member_accessor.py:
```diff
@@ -128,6 +151,15 @@ def get_submodule(self, name: str) -> "torch.nn.Module":
+        ``accessor.swap_submodule("layer1.conv1", conv2)``.
+        """
+        prefix, _, attr = path.rpartition(".")
+        return swap_submodule(self.get_submodule(prefix), attr, value)
```
Swapping submodules may break the caching mechanism in `NamedMemberAccessor`. If we want this works as expected, the batched version of `swap_submodule` should handle the order carefully.

```python
accessor.swap_submodule('layer1.conv1', myconv1)
accessor.swap_submodule('layer1', mylayer1)
```

```python
accessor.swap_submodule('layer1', mylayer1)
accessor.swap_submodule('layer1.conv1', myconv1)
```
